### PR TITLE
Parallelize tasks; enable early signaling connection establishment.

### PR DESCRIPTION
**Pull request by richnew10**
Opening the signaling websocket is one of the biggest sources of delay
when joining a meeting: even a fast connection to a nearby region can
take over 300ms.

This commit introduces an argument to `start` to allow a 'pre-start'
behavior to occur: call `start({ signalingOnly: true })` to begin signaling, and then call
`start()` to complete peer connection setup and media negotiation.

This change also slightly parallelizes some start tasks.

**Testing**

> 1. Have you successfully run `npm run build:release` locally?

Yes.

> 2. How did you test these changes?

With the meeting demo.

> 3. Can these changes be tested using the demo application? If yes, which demo application can be used to test it?

See above.

> 4. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved?

Yes: `AudioVideoFacade.start` now takes an options map as an argument, allowing a signaling-only start to occur.

The alternative is to add a new method to the facade and controller interfaces.

> 5. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved?

No, though this does cause the meeting app to establish a signaling connection earlier — in the lobby. Customer applications doing so will cause a small increase in load on the service.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/audiovideofacade/DefaultAudioVideoFacade.ts [hunk 1]
```diff
@@ -48,8 +48,8 @@ export default class DefaultAudioVideoFacade implements AudioVideoFacade {
     this.audioVideoController.setAudioProfile(audioProfile);
   }
 
-  start(): void {
-    this.audioVideoController.start();
+  start(options?: { signalingOnly?: boolean }): void {
```
Comment (aburnson) [inline src/audiovideofacade/DefaultAudioVideoFacade.ts:51]: Benefits of this approach over adding an explicit top-level `prestart()` method?

Comment (richnew10) [reply src/audiovideofacade/DefaultAudioVideoFacade.ts:51]: Firstly, it avoids creating an optional method on a core interface -- a customer's `AudioVideoController` subclass doesn't need to change even if the facade does, and that's true even without having to check `if (this.audioVideoController.prestart) { ... }`.

Secondly, this is a one-to-two transition, from "just start" to "maybe there are several kinds or phases of starting", and so I'm looking ahead to the next: check-in mode, full connection but don't start sending media yet, whatever else pops up.

I don't feel strongly about it, and you have both deeper knowledge of the signaling flow and longer experience with the design intent of this API, so happy to receive guidance.

File: src/audiovideocontroller/DefaultAudioVideoController.ts [hunk 2]
```diff
@@ -359,47 +390,53 @@ export default class DefaultAudioVideoController
...
+                new ListenForVolumeIndicatorsTask(this.meetingSessionContext),
+                new SendAndReceiveDataMessagesTask(this.meetingSessionContext),
+                new JoinAndReceiveIndexTask(this.meetingSessionContext),
+                new ReceiveTURNCredentialsTask(this.meetingSessionContext),
+                // TODO: ensure index handler does not race with incoming index update
+                new ReceiveVideoStreamIndexTask(this.meetingSessionContext),
+              ]),
+              new SerialGroupTask(this.logger, 'Peer', [
+                new CreatePeerConnectionTask(this.meetingSessionContext),
+                new AttachMediaInputTask(this.meetingSessionContext),
```
[the remark this answers is not in the document]
Comment (richnew10) [reply src/audiovideocontroller/DefaultAudioVideoController.ts:420]: In an ambitious alternative future we rework all of these tasks and all of the fields they use (e.g., the signaling client itself) to be a tree of promises, allowing the dependency tree and timeouts to control order of execution and parallelism. Maybe not today!

The approach I took first was to define `PromiseTask`, which just waits on an already-established promise. As you saw, I reworked that to `OnceTask`, allowing the same `Task` to be present multiple times in the task tree. We can use that same trick here and elsewhere by representing a resolves-before relationship by putting the dependent task earlier in the same serial group:

```typescript
  new SerialGroupTask(this.logger, 'Peer', [
    new CreatePeerConnectionTask(this.meetingSessionContext),
    previousReceiveAudioInputTask,
    new AttachMediaInputTask(this.meetingSessionContext),
```

This gives us a synchronization barrier without much work. Thoughts?

**Comment by richnew10** [comment]
Evaluating performance of this:

* By default, the only change will be to parallelize meeting join. The signaling change is behind a checkbox that is off by default. Specify `earlyConnect=1` in the URL to turn it on, or check it by hand.
* The entire connection flow is in a serial task labeled `AudioVideoStart`.
* If you check the checkbox, `OpenSignalingConnectionTask` will not contribute to `AudioVideoStart` in most cases.

So: compare the `AudioVideoStart took` values between this and a previous release, and then compare with `?earlyConnect=1` in the URL to see the preconnect improvement.

**Comment by richnew10** [comment]
Three failures to work on.

**Comment by richnew10** [comment]
Testing in Sydney from the west coast of the US:

```
Early join, promises:
AudioVideoStart took 976 ms
AudioVideoStart took 952 ms

No early join, promises:
AudioVideoStart took 1712 ms
AudioVideoStart took 1757 ms
AudioVideoStart took 1959 ms

No early join, task:
AudioVideoStart took 1745 ms
AudioVideoStart took 1912 ms
AudioVideoStart took 2011 ms
```

It looks like the promise-based approach in this Chrome fast path (permissions already granted etc.) shaves off 30–50msec, and the early signaling connection establishment saves that entire time, which is about a second to Sydney.

The long and the short of it: this change *halves* the time to enter the meeting for people in the most distant regions, and saves about 300ms for those close to their meeting region.

**Comment by richnew10** [comment]
N.B., I will write some additional guidance docs for this as a follow-up, but I didn't want to block review (and potential landing) on that.

I suggest flipping on the checkbox in the meeting demo for the majority of integration tests: it will speed up the tests in general, but leave a few to take the slow path.

File: src/audiovideocontroller/DefaultAudioVideoController.ts [hunk 3]
```diff
@@ -104,8 +105,12 @@ export default class DefaultAudioVideoController
   private enableSimulcast: boolean = false;
   private totalRetryCount = 0;
   private startAudioVideoTimestamp: number = 0;
+  private signalingTask: Task;
   destroyed = false;
 
+  /** @internal */
+  usePromises: boolean = true;
+
```
Comment (richnew10) [inline src/audiovideocontroller/DefaultAudioVideoController.ts:113]: Leaving this in for a little while to have an easier way to revert this change if needed.

File: src/task/NoOpTask.ts [hunk 4]
```diff
@@ -11,7 +11,7 @@ export default class NoOpTask implements Task {
   }
 
   run(): Promise<void> {
-    return;
+    return Promise.resolve();
   }
```
Comment (richnew10) [inline src/task/NoOpTask.ts:15]: This fix is the result of a disappointingly large amount of wasted time wondering why my tests were broken!

File: src/audiovideocontroller/DefaultAudioVideoController.ts [hunk 5]
```diff
@@ -242,6 +244,166 @@ export default class DefaultAudioVideoController
...
+                ]),
+                this.meetingSessionContext.meetingSessionConfiguration.attendeePresenceTimeoutMs
+              )
+            : /* istanbul ignore next */ new SetRemoteDescriptionTask(context),
+        ])
+      ),
+    ]);
+  }
+
+  private connectWithTasks(needsToWaitForAttendeePresence: boolean): Task {
```
Comment (richnew10) [inline src/audiovideocontroller/DefaultAudioVideoController.ts:360]: This is the old implementation. Will remove when we're confident.

File: src/task/OnceTask.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    if (this.promise) {
+      return this.promise;
+    }
+
+    const dependencies = this.dependencies
+      ? Promise.all(this.dependencies.map(d => d?.run()))
+      : Promise.resolve();
+
+    this.ongoing = dependencies.then(() => {
+      this.logDependencies();
```
Comment (devalevenkatesh) [inline src/task/OnceTask.ts:97]: Understanding: The `logDependencies` function will be executed once all dependency promises are resolved right ? as it is getting called in `then`?

Comment (richnew10) [reply src/task/OnceTask.ts:97]: Correct.

File: CHANGELOG.md [hunk 7]
```diff
@@ -9,12 +9,21 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- `DefaultAudioVideoFacade.start` now takes an options argument. You can use
+  this to trigger a signaling socket connection prior to device selection: call
+  `audioVideo.start({ signalingOnly: true })`, and then later call
+  `audioVideo.start()` as usual.
+
 ### Changed
 - `startVideoPreviewForVideoInput` uses the active video input stream instead of calling `getUserMedia` again
```
[the remark this answers is not in the document]
Comment (richnew10) [reply CHANGELOG.md:19]: It got rebased into the wrong place: should be one line higher.

File: src/task/OnceTask.ts (hunk 6 above)
Comment (devalevenkatesh) [reply src/task/OnceTask.ts:97]: So then the `logDependencies` function logs 'waiting for dependencies:..', should it be logging 'completed'?

Comment (richnew10) [reply src/task/OnceTask.ts]: Switched this to log after triggering the deps.

File: src/audiovideocontroller/DefaultAudioVideoController.ts [hunk 8]
```diff
@@ -242,6 +244,166 @@ export default class DefaultAudioVideoController
...
+      // Just-in-case error handling.
+      .catch(
+        /* istanbul ignore next */
+        e => {
+          this.logger.error(`start failed: ${e}`);
+        }
+      );
+  }
+
+  // For tests only.
```
Comment (ltrung) [inline src/audiovideocontroller/DefaultAudioVideoController.ts:275]: nit: Is this accurate since we called it under start()?

File: src/task/OnceTask.ts [hunk 9]
```diff
@@ -0,0 +1,109 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import Logger from '../logger/Logger';
+import LogLevel from '../logger/LogLevel';
+import Task from './Task';
+
+function nextTick(f: () => void): void {
```
Comment (ltrung) [inline src/task/OnceTask.ts:8]: Should we use AsyncScheduler here since you introduced nextTick there already?

File: integration/configs/app_quit_video_test.config.json [hunk 10]
```diff
@@ -10,7 +10,7 @@
       "noOfThreads": 1,
       "testImpl": "AppQuitVideoTest.js",
       "payload": {
-        "url": "http://localhost:8080/",
+        "url": "http://localhost:8080/?earlyConnect=1",
```
Comment (ltrung) [inline integration/configs/app_quit_video_test.config.json:13]: Is there any specific reason we enable this for these particular tests but not the others?

File: test/audiovideocontroller/DefaultAudioVideoController.test.ts [hunk 11]
```diff
@@ -333,6 +336,123 @@ describe('DefaultAudioVideoController', () => {
...
+      webSocketAdapter.send(makeAudioStreamIdInfoFrame());
+      webSocketAdapter.send(makeAudioMetadataFrame());
+
+      expect(sessionStarted).to.be.true;
+      expect(sessionConnecting).to.be.true;
+
+      await stop();
+    });
+
+    it('can be started with a pre-start', async () => {
```
Comment (ltrung) [inline test/audiovideocontroller/DefaultAudioVideoController.test.ts:391]: Do we need a test case for prestart and then stop and make sure we clean up everything correctly?

File: integration/configs/app_quit_video_test.config.json (hunk 10 above)
Comment (richnew10) [reply integration/configs/app_quit_video_test.config.json:13]: See earlier commentary — it makes sure we get some coverage on both paths without having to do something redundant and costly like running the entire suite twice.

**Review by ltrung — COMMENTED** [review]
Could we update the PR Description with how testing should be done?

File: src/task/OnceTask.ts (hunk 9 above)
Comment (richnew10) [reply src/task/OnceTask.ts:8]: Sure.

File: guides/03_API_Overview.md [hunk 12]
```diff
@@ -142,14 +142,14 @@ You may optionally listen to the following callbacks to monitor aspects of conne
...
 
-To start the meeting session, call meetingSession.audioVideo.[start()](https://aws.github.io/amazon-chime-sdk-js/interfaces/audiovideofacade.html#start). This method will initialize all underlying components, set up connections, and immediately start sending and receiving audio.
-
-To stop the meeting session, call meetingSession.audioVideo.[stop()](https://aws.github.io/amazon-chime-sdk-js/interfaces/audiovideofacade.html#stop).
+To stop the meeting session, call `meetingSession.audioVideo.[stop()](https://aws.github.io/amazon-chime-sdk-js/interfaces/audiovideofacade.html#stop)`.
 
 The `stop()` method does not clean up observers. You can start and stop a session multiple times using the same observers. In other words observers are not tied to the lifecycle of the session.
 
+You can specify a `signalingOnly` option when calling `start` to cause only the initial signaling connection to be established, then complete meeting join with a second call to `start` after choosing audio and video sources. This can improve meeting join latency.
```
Comment (ltrung) [inline guides/03_API_Overview.md:151]: When do builders want to do this? Could we give a more specific use case?

Comment (ltrung) [inline guides/03_API_Overview.md:151]: When do builders want to do this? Could we give a more specific use case?

File: integration/configs/app_quit_video_test.config.json (hunk 10 above)
Comment (ltrung) [reply integration/configs/app_quit_video_test.config.json:13]: Got it. I understand the reasoning just wonder if there is any specific reason to choose these specific tests :)

File: guides/03_API_Overview.md (hunk 12 above)
Comment (richnew10) [reply guides/03_API_Overview.md:151]: It's discussed a little in the FAQ.

File: src/audiovideocontroller/DefaultAudioVideoController.ts (hunk 8 above)
Comment (richnew10) [reply src/audiovideocontroller/DefaultAudioVideoController.ts:275]: The comment implies "this is public for tests only". I'll clarify 'cos evidently the implication isn't clear enough!

File: integration/configs/app_quit_video_test.config.json (hunk 10 above)
Comment (richnew10) [reply integration/configs/app_quit_video_test.config.json:13]: Nope, no reason. I just picked a few at random.

File: src/audiovideofacade/DefaultAudioVideoFacade.ts (hunk 1 above)
Comment (ltrung) [reply src/audiovideofacade/DefaultAudioVideoFacade.ts:51]: Do we see that any app would want to do this as best practice to reduce join time? If so, it doesn't feel like it should be an option.
Is there any plan to change this in future (maybe v3)?

File: test/audiovideocontroller/DefaultAudioVideoController.test.ts (hunk 11 above)
Comment (richnew10) [reply test/audiovideocontroller/DefaultAudioVideoController.test.ts:391]: I'll add one.

File: src/audiovideofacade/DefaultAudioVideoFacade.ts (hunk 1 above)
Comment (richnew10) [reply src/audiovideofacade/DefaultAudioVideoFacade.ts:51]: Most apps will want to, but knowing _when_ is up to the application.

File: guides/03_API_Overview.md (hunk 12 above)
Comment (richnew10) [reply guides/03_API_Overview.md]: Added an example.

**Review by aburnson — APPROVED** [review]

**Review by devalevenkatesh — APPROVED** [review]
